src/Microsoft.AspNetCore.Razor/Compilation/TagHelpers/TagHelperDescriptorProvider.cs

```diff
@@ -77,29 +76,34 @@ public TagHelperDescriptorProvider(IEnumerable<TagHelperDescriptor> descriptors)
                 descriptors = matchingDescriptors.Concat(descriptors);
             }
 
-            var applicableDescriptors = ApplyRequiredAttributes(descriptors, attributes);
-            applicableDescriptors = ApplyParentTagFilter(applicableDescriptors, parentTagName);
+            var applicableDescriptors = new List<TagHelperDescriptor>();
+            foreach (var descriptor in descriptors)
+            {
+                if (HasRequiredAttributes(descriptor, attributes) &&
+                    HasRequiredParentTag(descriptor, parentTagName))
+                {
+                    applicableDescriptors.Add(descriptor);
+                }
+            }
 
-            return applicableDescriptors.ToArray();
+            return applicableDescriptors;
         }
 
-        private IEnumerable<TagHelperDescriptor> ApplyParentTagFilter(
-            IEnumerable<TagHelperDescriptor> descriptors,
+        private bool HasRequiredParentTag(
+            TagHelperDescriptor descriptor,
             string parentTagName)
         {
-            return descriptors.Where(descriptor =>
-                descriptor.RequiredParent == null ||
-                string.Equals(parentTagName, descriptor.RequiredParent, StringComparison.OrdinalIgnoreCase));
+            return descriptor.RequiredParent == null ||
+                string.Equals(parentTagName, descriptor.RequiredParent, StringComparison.OrdinalIgnoreCase);
         }
 
-        private IEnumerable<TagHelperDescriptor> ApplyRequiredAttributes(
-            IEnumerable<TagHelperDescriptor> descriptors,
+        private bool HasRequiredAttributes(
+            TagHelperDescriptor descriptor,
             IEnumerable<KeyValuePair<string, string>> attributes)
         {
-            return descriptors.Where(
-                descriptor => descriptor.RequiredAttributes.All(
-                    requiredAttribute => attributes.Any(
-                        attribute => requiredAttribute.IsMatch(attribute.Key, attribute.Value))));
+            return descriptor.RequiredAttributes.All(
+                requiredAttribute => attributes.Any(
+                    attribute => requiredAttribute.IsMatch(attribute.Key, attribute.Value)));
```

Comment:
Unfortunately this is backed by an `IEnumerable` at times and there's not a lot we can do to fix this.